sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java

```diff
@@ -0,0 +1,475 @@
+import org.apache.calcite.rel.core.Sort;
+import org.apache.calcite.rel.core.TableModify;
+import org.apache.calcite.rel.core.TableScan;
+import org.apache.calcite.rel.core.Uncollect;
+import org.apache.calcite.rel.core.Window;
+import org.apache.calcite.rel.stream.Delta;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.util.Pair;
+
+import org.apache.hadoop.classification.InterfaceStability;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.DAG;
+import com.datatorrent.api.Operator;
+import com.datatorrent.lib.io.ConsoleOutputOperator;
+
+/**
+ * This class is the main class that converts relational algebra to a sub-DAG.
+ */
+@InterfaceStability.Evolving
+public class RelNodeVisitor
+{
+  private final DAG dag;
+  private final TupleSchemaRegistry tupleSchemaRegistry;
+  private final JavaTypeFactory typeFactory;
+
+  public RelNodeVisitor(DAG dag, JavaTypeFactory typeFactory)
+  {
+    this.dag = dag;
+    this.typeFactory = typeFactory;
+    this.tupleSchemaRegistry = new TupleSchemaRegistry();
+  }
+
+  private static Class convertPrimitiveToSqlType(RelDataType type)
+  {
+    /* I hope that following this method instead of calling value.value() is better
+    because we can catch any type mismatches. */
+    switch (type.getSqlTypeName()) {
+      case BOOLEAN:
+        return Boolean.class;
+      case TINYINT:
+      case SMALLINT:
+      case INTEGER:
+        return Integer.class;
+      case BIGINT:
+        return Long.class;
+      case REAL:
+        return Float.class;
+      case FLOAT:
+      case DOUBLE:
+        return Double.class;
+      case DATE:
+        return Date.class;
+      case TIME:
+        return Date.class;
+      case TIMESTAMP:
+        return Date.class;
+      case CHAR:
+      case VARCHAR:
+        return String.class;
+      case BINARY:
+      case VARBINARY:
+        return Byte.class;
+      case ANY:
+      case SYMBOL:
+        return Object.class;
+      default:
+        throw new RuntimeException(String.format("Unsupported type %s", type.getSqlTypeName()));
+    }
+  }
+
+  /**
+   * This is the main method in this relational node visitor which traverses the relational algebra in reverse direction
+   * and populate the given underlying DAG object.
+   *
+   * @param relNode RelNode which needs to be traversed.
+   *
+   * @return RelInfo representing information of current stage
+   * @throws Exception
+   */
+  public final RelInfo traverse(RelNode relNode) throws Exception
+  {
+    List<RelInfo> inputStreams = new ArrayList<>();
+    for (RelNode input : relNode.getInputs()) {
+      inputStreams.add(traverse(input));
+    }
+
+    RelInfo currentNodeRelInfo;
+
+    if (relNode instanceof Aggregate) {
```

Comment:
Could we use a map relNode.getClass -> visitor/translator with a common interface instead of the if-then-else? That would also allow for override of visitors for customization.